crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -364,28 +389,150 @@ fn state_cleaner<T: Clone + Resource>(mut state: ResMut<State<T>>) -> ShouldRun
+            )
+            .add_system_set(
+                SystemSet::default()
+                    .with_system(
+                        (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                            r.push("update S4");
+                            s.overwrite_push(MyState::S5).unwrap();
+                        })
+                        .system(),
+                    )
+                    .with_run_criteria(State::<MyState>::on_update(MyState::S4)),
+            )
+            .add_system_set(
+                SystemSet::default()
+                    .with_system(
+                        (|mut r: ResMut<Vec<&'static str>>| r.push("inactive S4")).system(),
+                    )
+                    .with_run_criteria(State::<MyState>::on_inactive_update(MyState::S4)),
+            )
+            .add_system_set(
+                SystemSet::default()
+                    .with_system(
+                        (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                            r.push("update S5");
+                            s.overwrite_push(MyState::S6).unwrap();
+                        })
+                        .system(),
+                    )
+                    .with_run_criteria(State::<MyState>::on_update(MyState::S5)),
+            )
+            .add_system_set(
+                SystemSet::default()
+                    .with_system(
+                        (|mut r: ResMut<Vec<&'static str>>| r.push("inactive S5")).system(),
+                    )
+                    .with_run_criteria(State::<MyState>::on_inactive_update(MyState::S5)),
+            )
+            .add_system_set(
+                SystemSet::default()
+                    .with_system(
+                        (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                            r.push("update S6");
+                            s.overwrite_push(MyState::Final).unwrap();
+                        })
+                        .system(),
+                    )
+                    .with_run_criteria(State::<MyState>::on_update(MyState::S6)),
+            )
+            .add_system_set(
+                SystemSet::default()
+                    .with_system((|mut r: ResMut<Vec<&'static str>>| r.push("resume S4")).system())
+                    .with_run_criteria(State::<MyState>::on_resume(MyState::S4)),
+            )
+            .add_system_set(
+                SystemSet::default()
+                    .with_system((|mut r: ResMut<Vec<&'static str>>| r.push("exit S4")).system())
+                    .with_run_criteria(State::<MyState>::on_exit(MyState::S4)),
+            );
+
+        loop {
+            println!("new run!");
+            stage.run(&mut world, &mut resources);
+            for s in resources.get_mut::<Vec<&'static str>>().unwrap().drain(..) {
+                println!("{}", s);
```

Comment:
That's not how tests work. If you expect a known output, compare what you get to it.

Reply:
whoops, forgot that :P